File: dev/Layer.js
"use strict"

class Layer {
    
    constructor (size, importedData) {
        this.size = size
        this.neurons = [...new Array(size)].map((n, ni) => new Neuron(importedData ? importedData[ni] : undefined))
        this.state = "not-initialised"
    }

    assignNext (layer) {
        this.nextLayer = layer
    }

    assignPrev (layer) {
        this.prevLayer = layer
        this.neurons.forEach(neuron => neuron.init(layer.size, {
            adaptiveLR: this.adaptiveLR,
            activationConfig: this.activationConfig,
            eluAlpha: this.eluAlpha
        }))
        this.state = "initialised"
    }

    forward (data) {

        this.neurons.forEach((neuron, ni) => {

            if(this.state=="training" && (neuron.dropped = Math.random() > this.dropout)) {
                neuron.activation = 0
            }else {
                neuron.sum = neuron.bias
                this.prevLayer.neurons.forEach((pNeuron, pni) => neuron.sum += pNeuron.activation * neuron.weights[pni])
                neuron.activation = this.activation(neuron.sum, false, neuron) / (this.dropout|1)
            }
        })
    }

    backward (expected) {
        this.neurons.forEach((neuron, ni) => {

            if(neuron.dropped) {
                neuron.error = 0
                neuron.deltaBias = 0
            }else {
                if(typeof expected !== "undefined") {
                    neuron.error = expected[ni] - neuron.activation
                }else {
                    neuron.derivative = this.activation(neuron.sum, true, neuron)
                    neuron.error = neuron.derivative * this.nextLayer.neurons.map(n => n.error * (n.weights[ni]|0))
                                                                             .reduce((p,c) => p+c, 0)
                }

                neuron.weights.forEach((weight, wi) => {
                    neuron.deltaWeights[wi] += (neuron.error * this.prevLayer.neurons[wi].activation) * 
                                               (1 + (this.l2||0) * neuron.deltaWeights[wi])
                })

                neuron.deltaBias = neuron.error
            }            
        })
    }
}

typeof window=="undefined" && (global.Layer = Layer) 

File: dev/cpp/Layer.cpp

Layer::Layer(int netI, int s) {
    netInstance = netI;
    size = s;
}

Layer::~Layer () {
    for (int n=0; n<neurons.size(); n++) {
        delete neurons[n];
    }
}

void Layer::assignNext (Layer* l) {
    nextLayer = l;
}

void Layer::assignPrev (Layer* l) {
    prevLayer = l;
}

void Layer::init (int layerIndex) {
    for (int n=0; n<size; n++) {
        Neuron* neuron = new Neuron();

        if (layerIndex) {

            for (int pn=0; pn<prevLayer->size; pn++) {
                // neuron->weights.push_back(rand()/1e+11-);
                neuron->weights.push_back(((double) rand() / (RAND_MAX))/5 - 0.1);
            }

            neuron->bias = ((double) rand() / (RAND_MAX))/5 - 0.1;
        }

        neuron->init(netInstance);
        neurons.push_back(neuron);
    }
}

void Layer::forward (void) {
    for (int n=0; n<neurons.size(); n++) {
        neurons[n]->sum = neurons[n]->bias;

        for (int pn=0; pn<prevLayer->neurons.size(); pn++) {
            neurons[n]->sum += prevLayer->neurons[pn]->activation * neurons[n]->weights[pn];
        }

        neurons[n]->activation = activation(neurons[n]->sum, false, neurons[n]);
    }
}

void Layer::backward (std::vector<double> expected) {
    for (int n=0; n<neurons.size(); n++) {

        if (expected.size()) {
            neurons[n]->error = expected[n] - neurons[n]->activation;
        } else {
            neurons[n]->derivative = activation(neurons[n]->sum, true, neurons[n]);

            double weightedErrors = 0.0;

            for (int nn=0; nn<nextLayer->neurons.size(); nn++) {
                weightedErrors += nextLayer->neurons[nn]->error * nextLayer->neurons[nn]->weights[n];
            }

            neurons[n]->error = neurons[n]->derivative * weightedErrors;
        }

        for (int wi=0; wi<neurons[n]->weights.size(); wi++) {
            neurons[n]->deltaWeights[wi] += neurons[n]->error * prevLayer->neurons[wi]->activation; //* (1  neurons[n]->deltaWeights[dw]);
        }

        neurons[n]->deltaBias = neurons[n]->error;
    }
}

void Layer::applyDeltaWeights (void) {

    // Function pointers are far too slow, here.
    // Using code repetitive switch statements makes a substantial perf difference
    int updateFnIndex = Network::getInstance(netInstance)->updateFnIndex;

    switch (updateFnIndex) {
        case 0: // vanilla
            for(int n=0; n<neurons.size(); n++) {
                for (int dw=0; dw<neurons[n]->deltaWeights.size(); dw++) {
                    neurons[n]->weights[dw] = NetMath::vanillaupdatefn(netInstance, neurons[n]->weights[dw], neurons[n]->deltaWeights[dw]);
                }
                neurons[n]->bias = NetMath::vanillaupdatefn(netInstance, neurons[n]->bias, neurons[n]->deltaBias);
            }
            break;
        case 1: // gain
            for(int n=0; n<neurons.size(); n++) {
                for (int dw=0; dw<neurons[n]->deltaWeights.size(); dw++) {
                    neurons[n]->weights[dw] = NetMath::gain(netInstance, neurons[n]->weights[dw], neurons[n]->deltaWeights[dw], neurons[n], dw);
                }
                neurons[n]->bias = NetMath::gain(netInstance, neurons[n]->bias, neurons[n]->deltaBias, neurons[n], -1);
            }
            break;
        case 2: // adagrad
            for(int n=0; n<neurons.size(); n++) {
                for (int dw=0; dw<neurons[n]->deltaWeights.size(); dw++) {
                    neurons[n]->weights[dw] = NetMath::adagrad(netInstance, neurons[n]->weights[dw], neurons[n]->deltaWeights[dw], neurons[n], dw);
                }
                neurons[n]->bias = NetMath::adagrad(netInstance, neurons[n]->bias, neurons[n]->deltaBias, neurons[n], -1);
            }
            break;
        case 3: // rmsprop
            for(int n=0; n<neurons.size(); n++) {
                for (int dw=0; dw<neurons[n]->deltaWeights.size(); dw++) {
                    neurons[n]->weights[dw] = NetMath::rmsprop(netInstance, neurons[n]->weights[dw], neurons[n]->deltaWeights[dw], neurons[n], dw);
                }
                neurons[n]->bias = NetMath::rmsprop(netInstance, neurons[n]->bias, neurons[n]->deltaBias, neurons[n], -1);
            }
            break;
    }
}

void Layer::resetDeltaWeights (void) {
    for(int n=0; n<neurons.size(); n++) {
        for (int dw=0; dw<neurons[n]->deltaWeights.size(); dw++) {
            neurons[n]->deltaWeights[dw] = 0;
        }
    }
}
